File: src/python/infer.py
import argparse
import json
import numpy as np
import pandas as pd
import torch
import torch.nn as nn


T = 512
IN_CHANNELS = 6
OUT_CHANNELS = 6
CPL = 16
LAYER_COUNT = 7
K = 3
DILATIONS = [1, 2, 4, 8, 16, 32, 64]

DEVICE = "cuda" if torch.cuda.is_available() else "cpu"


class CausalConv1d(nn.Module):
    def __init__(self, in_ch, out_ch, kernel_size, dilation):
        super().__init__()
        self.pad = dilation * (kernel_size - 1)
        self.conv = nn.Conv1d(
            in_ch,
            out_ch,
            kernel_size,
            dilation=dilation,
            padding=0
        )

    def forward(self, x):
        x = nn.functional.pad(x, (self.pad, 0))
        return self.conv(x)


class ResidualBlock(nn.Module):
    def __init__(self, in_ch, out_ch, dilation, use_projection):
        super().__init__()

        self.conv1 = CausalConv1d(in_ch, out_ch, K, dilation)
        self.conv2 = CausalConv1d(out_ch, out_ch, K, dilation)

        if use_projection:
            self.residual = nn.Conv1d(in_ch, out_ch, kernel_size=1)
        else:
            self.residual = nn.Identity()

    def forward(self, x):
        res = self.residual(x)

        out = self.conv1(x)
        out = torch.relu(out)

        out = self.conv2(out)
        out = torch.relu(out)

        out = out + res
        return out


class StreamingTCN(nn.Module):
    def __init__(self):
        super().__init__()

        self.input_layer = ResidualBlock(
            in_ch=IN_CHANNELS,
            out_ch=CPL,
            dilation=DILATIONS[0],
            use_projection=True
        )

        self.hidden_layers = nn.ModuleList([
            ResidualBlock(
                in_ch=CPL,
                out_ch=CPL,
                dilation=DILATIONS[i],
                use_projection=False
            )
            for i in range(1, LAYER_COUNT)
        ])

        self.output_layer = nn.Conv1d(CPL, OUT_CHANNELS, kernel_size=1)

    def forward(self, x):
        x = self.input_layer(x)

        for layer in self.hidden_layers:
            x = layer(x)

        y = self.output_layer(x)
        return y[:, :, -1]


def load_imu_csv(path):
    df = pd.read_csv(path, header=None)

    # acc_x, acc_y, acc_z, gyro_x, gyro_y, gyro_z
    return df.iloc[:, [10, 11, 12, 4, 5, 6]].values.astype(np.float32)


def load_json_weights(model, weights_path):
    with open(weights_path, "r") as f:
        w = json.load(f)

    model.input_layer.conv1.conv.weight.data = torch.tensor(w["inputFilter1"], dtype=torch.float32)
    model.input_layer.conv1.conv.bias.data = torch.tensor(w["inputBias1"], dtype=torch.float32)

    model.input_layer.conv2.conv.weight.data = torch.tensor(w["inputFilter2"], dtype=torch.float32)
    model.input_layer.conv2.conv.bias.data = torch.tensor(w["inputBias2"], dtype=torch.float32)

    model.input_layer.residual.weight.data = torch.tensor(w["inputResidualFilter"], dtype=torch.float32).unsqueeze(-1)
    model.input_layer.residual.bias.data = torch.tensor(w["inputResidualBias"], dtype=torch.float32)

    for i, layer in enumerate(model.hidden_layers):
        layer.conv1.conv.weight.data = torch.tensor(w["hiddenFilter1"][i], dtype=torch.float32)
        layer.conv1.conv.bias.data = torch.tensor(w["hiddenBias1"][i], dtype=torch.float32)

        layer.conv2.conv.weight.data = torch.tensor(w["hiddenFilter2"][i], dtype=torch.float32)
        layer.conv2.conv.bias.data = torch.tensor(w["hiddenBias2"][i], dtype=torch.float32)

    model.output_layer.weight.data = torch.tensor(w["outputFilter"], dtype=torch.float32).unsqueeze(-1)
    model.output_layer.bias.data = torch.tensor(w["outputBiases"], dtype=torch.float32)

    mean = np.array(w["mean"], dtype=np.float32)
    std = np.array(w["std"], dtype=np.float32)

    return mean, std


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--csv", required=True)
    parser.add_argument("--weights", required=True)
    parser.add_argument("--out", default="pytorch_reference.csv")
    args = parser.parse_args()

    model = StreamingTCN().to(DEVICE)
    mean, std = load_json_weights(model, args.weights)
    model.eval()

    data = load_imu_csv(args.csv)
    data = (data - mean) / std

    with open(args.out, "w") as f:
        f.write("sample,pred0,pred1,pred2,pred3,pred4,pred5\n")

        with torch.no_grad():
            for sample in range(T, len(data)):
                x = data[sample - T + 1: sample + 1]
                x = torch.tensor(x.T, dtype=torch.float32).unsqueeze(0).to(DEVICE)

                pred = model(x).cpu().numpy()[0]

                f.write(str(sample))
                for ch in range(OUT_CHANNELS):
                    f.write(f",{pred[ch]:.10f}")
                f.write("\n")

    print(f"Saved PyTorch predictions to {args.out}")


if __name__ == "__main__":
    main()

File: src/cpp/infer.cpp
/*
Model Architecture

Input: 6 channels
Layers: 7
Channels per layer: 16
Kernel size: 3
Dilation: 1,2,4,8,16,32,64
Output: 6 channels (prediction)
Mode: streaming (latest point only)
*/
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <array>
#include <string>
#include <iomanip>
#include "nlohmann/json.hpp"

using namespace std;
using json = nlohmann::json;

const int inChannels = 6;
const int layerCount = 7;
const int cpl = 16;
const int k = 3;
const int dilations[layerCount] = {1, 2, 4, 8, 16, 32, 64};
const int outChannels = 6;
const int T = 512; // chosen to cover RF = 1 + 2 * (k - 1) * sum(dilations) = 509

double meanVals[inChannels];
double stdVals[inChannels];

double input[inChannels][T];
double midLayers[layerCount][cpl][T];
double hiddenLayers[layerCount][cpl][T];

double inputFilter1[cpl][inChannels][k];
double inputBias1[cpl];

double inputFilter2[cpl][cpl][k];
double inputBias2[cpl];

double inputResidualFilter[cpl][inChannels];
double inputResidualBias[cpl];

double hiddenFilter1[layerCount - 1][cpl][cpl][k];
double hiddenBias1[layerCount - 1][cpl];

double hiddenFilter2[layerCount - 1][cpl][cpl][k];
double hiddenBias2[layerCount - 1][cpl];

double outputFilter[outChannels][cpl];
double outputBiases[outChannels];

int head = T - 1;
int samplesSeen = 0;

void ReLU(double &x)
{
    if (x < 0)
        x = 0;
}

void buildIdx(int idx[k], int d)
{
    for (int i = 0; i < k; i++)
    {
        int delay = d * i;
        idx[i] = (delay < samplesSeen) ? (head - delay + T) & (T - 1) : -1; // bitmask wrap around works since 512 is a power of 2
    }
}

void dilatedConv(
    double *in,
    double *kernel,
    double &acc,
    const int idx[k])
{
    for (int i = 0; i < k; i++)
    {
        if (idx[i] < 0)
            break;
        acc += kernel[k - 1 - i] * in[idx[i]];
    }
}

void doInputLayer()
{
    int latest = head;
    int d = dilations[0];

    int idx[k];
    buildIdx(idx, d);

    // Conv1: input 6 -> 16
    for (int oc = 0; oc < cpl; oc++)
    {
        midLayers[0][oc][latest] = inputBias1[oc];

        for (int ic = 0; ic < inChannels; ic++)
            dilatedConv(input[ic], inputFilter1[oc][ic], midLayers[0][oc][latest], idx);

        ReLU(midLayers[0][oc][latest]);
    }

    // Conv2: 16 -> 16
    for (int oc = 0; oc < cpl; oc++)
    {
        hiddenLayers[0][oc][latest] = inputBias2[oc];

        for (int ic = 0; ic < cpl; ic++)
            dilatedConv(midLayers[0][ic], inputFilter2[oc][ic], hiddenLayers[0][oc][latest], idx);

        ReLU(hiddenLayers[0][oc][latest]);

        // Residual projection: input 6 -> 16
        double res = inputResidualBias[oc];

        for (int ic = 0; ic < inChannels; ic++)
            res += inputResidualFilter[oc][ic] * input[ic][latest];

        hiddenLayers[0][oc][latest] += res;
    }
}

void doHiddenLayer(int layerNum)
{
    int latest = head;
    int d = dilations[layerNum];
    int f = layerNum - 1;

    int idx[k];
    buildIdx(idx, d);

    // Conv1: 16 -> 16
    for (int oc = 0; oc < cpl; oc++)
    {
        midLayers[layerNum][oc][latest] = hiddenBias1[f][oc];

        for (int ic = 0; ic < cpl; ic++)
            dilatedConv(hiddenLayers[layerNum - 1][ic], hiddenFilter1[f][oc][ic], midLayers[layerNum][oc][latest], idx);

        ReLU(midLayers[layerNum][oc][latest]);
    }

    // Conv2: 16 -> 16
    for (int oc = 0; oc < cpl; oc++)
    {
        hiddenLayers[layerNum][oc][latest] = hiddenBias2[f][oc];

        for (int ic = 0; ic < cpl; ic++)
            dilatedConv(midLayers[layerNum][ic], hiddenFilter2[f][oc][ic], hiddenLayers[layerNum][oc][latest], idx);

        ReLU(hiddenLayers[layerNum][oc][latest]);

        // Identity residual: 16 -> 16
        hiddenLayers[layerNum][oc][latest] += hiddenLayers[layerNum - 1][oc][latest];
    }
}

void doOutputLayer(double output[outChannels])
{
    int latest = head;

    for (int oc = 0; oc < outChannels; oc++)
    {
        output[oc] = outputBiases[oc];

        for (int ic = 0; ic < cpl; ic++)
            output[oc] += outputFilter[oc][ic] * hiddenLayers[layerCount - 1][ic][latest];
    }
}

void shiftInput(double newSample[inChannels])
{
    head = (head + 1) & (T - 1);

    if (samplesSeen < T)
        samplesSeen++;

    for (int ch = 0; ch < inChannels; ch++)
        input[ch][head] = newSample[ch];
}

void inferNext(double newSample[inChannels], double prediction[outChannels])
{
    shiftInput(newSample);
    doInputLayer();

    for (int layer = 1; layer < layerCount; layer++)
        doHiddenLayer(layer);

    doOutputLayer(prediction);
}

double computeError(double prediction[outChannels], double actual[outChannels])
{
    double err = 0.0;

    for (int i = 0; i < outChannels; i++)
    {
        double diff = actual[i] - prediction[i];
        err += diff * diff;
    }

    return err;
}

// the inference math stops here, the following is for weight loading and running the inference

void loadWeights(const string& path)
{
    ifstream f(path);
    json j;
    f >> j;

    for (int i = 0; i < inChannels; i++) {
        meanVals[i] = j["mean"][i];
        stdVals[i] = j["std"][i];
    }

    for (int oc = 0; oc < cpl; oc++) {
        inputBias1[oc] = j["inputBias1"][oc];
        inputBias2[oc] = j["inputBias2"][oc];
        inputResidualBias[oc] = j["inputResidualBias"][oc];

        for (int ic = 0; ic < inChannels; ic++) {
            inputResidualFilter[oc][ic] = j["inputResidualFilter"][oc][ic];
            for (int kk = 0; kk < k; kk++)
                inputFilter1[oc][ic][kk] = j["inputFilter1"][oc][ic][kk];
        }

        for (int ic = 0; ic < cpl; ic++)
            for (int kk = 0; kk < k; kk++)
                inputFilter2[oc][ic][kk] = j["inputFilter2"][oc][ic][kk];
    }

    for (int l = 0; l < layerCount - 1; l++)
        for (int oc = 0; oc < cpl; oc++) {
            hiddenBias1[l][oc] = j["hiddenBias1"][l][oc];
            hiddenBias2[l][oc] = j["hiddenBias2"][l][oc];

            for (int ic = 0; ic < cpl; ic++)
                for (int kk = 0; kk < k; kk++) {
                    hiddenFilter1[l][oc][ic][kk] = j["hiddenFilter1"][l][oc][ic][kk];
                    hiddenFilter2[l][oc][ic][kk] = j["hiddenFilter2"][l][oc][ic][kk];
                }
        }

    for (int oc = 0; oc < outChannels; oc++) {
        outputBiases[oc] = j["outputBiases"][oc];
        for (int ic = 0; ic < cpl; ic++)
            outputFilter[oc][ic] = j["outputFilter"][oc][ic];
    }
}

int main(int argc, char* argv[])
{
    loadWeights(argv[2]);

    ifstream in(argv[1]);
    ofstream out(argv[3]);

    out << "sample,pred0,pred1,pred2,pred3,pred4,pred5\n";
    out << fixed << setprecision(10);

    string line;
    int sample = 0;

    while (getline(in, line)) {
        stringstream ss(line);
        string cell;
        vector<double> c;

        while (getline(ss, cell, ','))
            c.push_back(stod(cell));

        double x[inChannels] = {
            (c[10] - meanVals[0]) / stdVals[0],
            (c[11] - meanVals[1]) / stdVals[1],
            (c[12] - meanVals[2]) / stdVals[2],
            (c[4]  - meanVals[3]) / stdVals[3],
            (c[5]  - meanVals[4]) / stdVals[4],
            (c[6]  - meanVals[5]) / stdVals[5]
        };

        double pred[outChannels];
        inferNext(x, pred);

        if (sample >= T) {
            out << sample;
            for (int ch = 0; ch < outChannels; ch++)
                out << "," << pred[ch];
            out << "\n";
        }

        sample++;
    }

    return 0;
}